### DISCUSSION TITLE
Convergence and mechanism files

### DISCUSSION THREAD

**Discussion by JayStandridge** [AMReX-Combustion/PeleC#618] [Q&A] [answered]
Hi, I have a couple of questions:

1). It doesn't appear that PeleC tracks the solution residuals. If I am looking to obtain a steady state solution is there a way to see how much the solution changes each iteration?

2). Right now, I am trying to access some of the functions defined in LiDryer/mechanism.cpp. These are declared in mechanism.H, but it doesn't look like the mechanism.cpp file is linked. My code compiles, but fails to link. Specifically, I want the CKINU function, or some other way to get access to the stoichiometric coefficients in a reaction-independent manner- where should I look?

3). The PMF case takes in a dat file when phi is negative. The dat file specifies a 1-D solution to the problem which is then perturbed for the initial value of the PMF case. I was wondering how this file is generated. If I wanted to create a different phi or a different mechanism, how would I go about generating the dat file?

Thank you

**Comment by marchdf** [comment, marked as answer]
For 1) you can look at the integrated quantities, either logged to stdout :
```
TIME= 5.167702024e-09 MASS        = 0.0002667415295
TIME= 5.167702024e-09 XMOM        = 1.526351211e-26
TIME= 5.167702024e-09 YMOM        = 3.084407797e-21
TIME= 5.167702024e-09 ZMOM        = 0.01137956977
TIME= 5.167702024e-09 RHO*e       = -497092.0649
TIME= 5.167702024e-09 RHO*K       = 0.3959849419
TIME= 5.167702024e-09 RHO*E       = -497091.6689
TIME= 5.167702024e-09 ENSTROPHY   = 0.1780262139
TIME= 5.167702024e-09 FUEL PROD   = 0
TIME= 5.167702024e-09 TEMP        = 474.003162
```

or to a specific file such as `datlog`, controlled via `amr.data_log   = datlog`, see TG input files for example.

**Reply by marchdf** [reply]
For 2), I can replicate this problem and am looking into it.

**Reply by marchdf** [reply]
For 3), @esclapez or @drummerdoc  probably has the best answer

**Reply by marchdf** [reply]
Ah so 2) is a legit problem that I need to fix in PelePhysics.

**Reply by marchdf** [reply]
2) will be fixed in https://github.com/AMReX-Combustion/PelePhysics/pull/382

**Comment by esclapez** [comment]
for 3), the PMF is generated from Cantera. You can add the following to your Cantera script (assuming `f` is your FreeFlame object):

```
# Store Pele PMF format
    csv_file = str(pref)+"_Phi"+str(phi[0])+'_X_CGS.csv'
    with open(csv_file, 'w') as outfile:
        writer = csv.writer(outfile,delimiter=' ',quotechar=' ', quoting=csv.QUOTE_MINIMAL)
        writer.writerow(['VARIABLES = "X" "temp" "u" "rho"']+['"'+x+'"' for x in gas.species_names])
        writer.writerow(['ZONE I=' + str(len(f.grid)) + ' FORMAT=POINT' + ' SPECFORMAT=MOLE'])
        for kk in range(len(f.grid)):
            f.set_gas_state(kk)
            writer.writerow([f.grid[kk]*100.0, gas.T, f.velocity[kk]*100.0, gas.density*1e-3]+list(gas.X))
```

**Reply by drummerdoc** [reply]
This sure seems like something that we ought to put into a util folder or something, doesn't it?

**Reply by baperry2** [reply]
I believe that this already exists! Although a bit of documentation could be helpful, I can add some

https://github.com/AMReX-Combustion/PelePhysics/blob/development/Utility/FF-CANTERA_wPMF.py

**Reply by drummerdoc** [reply]
I knew this topic seems familiar!
